File: frontend/src/api/segments.ts
import { getData, postData } from './http'

export interface ClusterRow {
  cluster_id: number
  n?: number
  share?: number
  conversion_rate?: number
  conversion_note?: string
  profile_means?: Record<string, number | null>
}

export interface SegmentsData {
  method?: string
  n_clusters?: number
  n_samples?: number
  disclaimer?: string
  clusters: ClusterRow[]
  feature_columns?: string[]
  label_excluded?: boolean
}

export function fetchSegments() {
  return getData<SegmentsData>('/segments')
}

export function assignSegment(body: { customer_id?: number; features?: Record<string, unknown> }) {
  return postData<{
    cluster_id: number
    distance?: number
    customer_id?: number
    cluster?: ClusterRow
    disclaimer?: string
  }>('/segments/assign', body)
}

// ---------------------------------------------------------------------------
// 阶段9：分群对比 / 投影
// ---------------------------------------------------------------------------

export interface SegmentCompareRow {
  algo: string
  k: number
  silhouette?: number
  calinski_harabasz?: number
  bic?: number
  error?: string
}

export interface SegmentCompareData {
  kmeans_k: number
  comparison: SegmentCompareRow[]
  stability: {
    algo: string
    k: number
    n_boot: number
    ari_mean: number
    ari_std: number
    note?: string
  }
  auto_names?: Record<string, string>
  disclaimer?: string
}

export interface SegmentProjectionData {
  method: string
  explained_variance: number[]
  n_points: number
  points: Array<{ x: number; y: number; cluster: number; customer_id?: number }>
  disclaimer?: string
}

export function fetchSegmentCompare() {
  return getData<SegmentCompareData>('/segments/compare')
}

export function fetchSegmentProjection() {
  return getData<SegmentProjectionData>('/segments/projection')
}

File: src/digital_marketing/schemas/segments.py
"""分群 DTO。"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class SegmentsData(BaseModel):
    method: str | None = None
    n_clusters: int | None = None
    n_samples: int | None = None
    disclaimer: str | None = None
    clusters: list[dict[str, Any]] = Field(default_factory=list)
    feature_columns: list[str] = Field(default_factory=list)
    label_excluded: bool | None = True


class AssignRequest(BaseModel):
    customer_id: int | None = None
    features: dict[str, Any] | None = None


class AssignData(BaseModel):
    cluster_id: int
    distance: float | None = None
    customer_id: int | None = None
    cluster: dict[str, Any] | None = None
    disclaimer: str | None = None
